File: java/tests/java8/sodium/MemoryTest1.java
package sodium;

import sodium.*;

public class MemoryTest1
{
    public static void main(String[] args)
    {
        new Thread() {
            public void run()
            {
                try {
                    while (true) {
                        System.out.println("memory "+Runtime.getRuntime().totalMemory());
                        Thread.sleep(5000);
                    }
                }
                catch (InterruptedException e) {
                    System.out.println(e.toString());
                }
            }
        }.start();

        EventSink<Integer> et = new EventSink<Integer>();
        Behavior<Integer> t = et.hold(0);
        Event<Integer> etens = et.map(x -> x/10);
        Event<Integer> changeTens = et.snapshot(t, (neu, old) ->
            neu.equals(old) ? null : neu).filterNotNull();
        Behavior<Behavior<Tuple2<Integer,Integer>>> oout =
            changeTens.map(tens -> t.map(tt -> new Tuple2<Integer,Integer>(tens, tt))).
            hold(t.map(tt -> new Tuple2<Integer,Integer>(0, tt)));
        Behavior<Tuple2<Integer,Integer>> out = Behavior.switchB(oout);
        Listener l = out.value().listen(tu -> {
            //System.out.println(tu.a+","+tu.b);
        });
        int i = 0;
        while (i < 1000000000) {
            et.send(i);
            i++;
        }
        l.unlisten();
    }
}

File: clojure/sodium/src/sodium/memory_test_1.clj
(ns sodium.memory-test-1
  (:require [sodium.core :refer :all])
  (:import [nz.sodium])
  (:import [nz.sodium Operational Stream StreamSink StreamLoop Tuple2 Cell CellSink Transaction Node])
  (:import [java.util Optional])
  (:gen-class))



(defn -main
  "Memory test #1"
  [& args]
  (println "Starting!")
  (.start (Thread. (fn []
                     (loop []
                       (println (str "memory " (.totalMemory (Runtime/getRuntime))))
                       (Thread/sleep 5000)
                       (recur)))))
  (let [et (StreamSink.)
        t (.hold et 0)
        etens (.map et (apply1 [x] (int (/ x 10))))
        change-tens (Stream/filterOptional (.snapshot et t (apply2 [neu old]
                                                                  (if (= neu old)
                                                                    (oe)
                                                                    (oo neu)))))
        oout (.hold (.map change-tens (apply1 [tens]
                                             (.map t (apply1 [tt]
                                                             (Tuple2. tens tt)))))
                    (.map t (apply1 [tt]
                                    (Tuple2. 0 tt))))
        out (Cell/switchC oout)
        l (.listen out (handler [tu] (comment println (str (.a tu) "," (.b tu)))))]

    (loop [i 0]
      (when (< i 1000000000)
        (.send et i)
        (recur (inc i))))
    (.unlisten l))
  (System/exit 0))
  
                     

